### PR TITLE
Studying Arrow Convertions

### PR DESCRIPTION
**Pull request by trickarcher**
@jpivarski
Use: `g++ -c -fPIC arrow_c_interface.cpp -o foo.o`
`g++ -shared -Wl,-soname,libfoo.so -o libfoo.so foo.o`
to build the c++ library. `use_arrow.py` is the python file. So far I can get the bytes from the C struct but not from the Arrow Array, it doesn't consider `pyarrow.Array` as a byte object.

### CODE REVIEW

**Comment by jpivarski** [comment]
I'm here. I'll take a look soonish. In the meantime, if you want to write some explanations of what you've tried, that will help me when I'm ready to look at it.

Thanks!

**Comment by trickarcher** [comment]
Yup, no problem. I'll do that in a jiffy!

**Comment by trickarcher** [comment]
Should be a little better now!

**Comment by jpivarski** [comment]
So, following [this page](https://arrow.apache.org/docs/python/ipc.html), the way to serialize data in pyarrow is:

```python
>>> import pyarrow
>>> import numpy
>>> pyarrow.array([0, 1, 2, 3, 4, 5])
<pyarrow.lib.Int64Array object at 0x7fed1723cf30>
[
  0,
  1,
  2,
  3,
  4,
  5
]
>>> batch = pyarrow.RecordBatch.from_arrays([pyarrow.array([0, 1, 2, 3, 4, 5])], ["one"])
>>> sink = pyarrow.BufferOutputStream()
>>> writer = pyarrow.RecordBatchStreamWriter(sink, batch.schema)
>>> writer.write_batch(batch)
>>> writer.close()
>>> buf = sink.getvalue()
>>> buf
<pyarrow.lib.Buffer object at 0x7fed144a7970>
>>> numpy.frombuffer(buf, dtype=numpy.uint8)
array([255, 255, 255, 255, 120,   0,   0,   0,  16,   0,   0,   0,   0,
         0,  10,   0,  12,   0,   6,   0,   5,   0,   8,   0,  10,   0,
         0,   0,   0,   1,   3,   0,  12,   0,   0,   0,   8,   0,   8,
         0,   0,   0,   4,   0,   8,   0,   0,   0,   4,   0,   0,   0,
         1,   0,   0,   0,  20,   0,   0,   0,  16,   0,  20,   0,   8,
         0,   6,   0,   7,   0,  12,   0,   0,   0,  16,   0,  16,   0,
         0,   0,   0,   0,   1,   2,  36,   0,   0,   0,  20,   0,   0,
         0,   4,   0,   0,   0,   0,   0,   0,   0,   8,   0,  12,   0,
         8,   0,   7,   0,   8,   0,   0,   0,   0,   0,   0,   1,  64,
         0,   0,   0,   3,   0,   0,   0, 111, 110, 101,   0, 255, 255,
       255, 255, 136,   0,   0,   0,  20,   0,   0,   0,   0,   0,   0,
         0,  12,   0,  22,   0,   6,   0,   5,   0,   8,   0,  12,   0,
        12,   0,   0,   0,   0,   3,   3,   0,  24,   0,   0,   0,  48,
         0,   0,   0,   0,   0,   0,   0,   0,   0,  10,   0,  24,   0,
        12,   0,   4,   0,   8,   0,  10,   0,   0,   0,  60,   0,   0,
         0,  16,   0,   0,   0,   6,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   2,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,  48,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   1,   0,   0,   0,   6,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   1,   0,   0,   0,   0,   0,
         0,   0,   2,   0,   0,   0,   0,   0,   0,   0,   3,   0,   0,
         0,   0,   0,   0,   0,   4,   0,   0,   0,   0,   0,   0,   0,
         5,   0,   0,   0,   0,   0,   0,   0, 255, 255, 255, 255,   0,
         0,   0,   0], dtype=uint8)
>>> numpy.frombuffer(buf, dtype=numpy.uint8).tostring()
b'\xff\xff\xff\xffx\x00\x00\x00\x10\x00\x00\x00\x00\x00\n\x00\x0c\x00\x06\x00\x05\x00\x08\x00\n\x00\x00\x00\x00\x01\x03\x00\x0c\x00\x00\x00\x08\x00\x08\x00\x00\x00\x04\x00\x08\x00\x00\x00\x04\x00\x00\x00\x01\x00\x00\x00\x14\x00\x00\x00\x10\x00\x14\x00\x08\x00\x06\x00\x07\x00\x0c\x00\x00\x00\x10\x00\x10\x00\x00\x00\x00\x00\x01\x02$\x00\x00\x00\x14\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x08\x00\x0c\x00\x08\x00\x07\x00\x08\x00\x00\x00\x00\x00\x00\x01@\x00\x00\x00\x03\x00\x00\x00one\x00\xff\xff\xff\xff\x88\x00\x00\x00\x14\x00\x00\x00\x00\x00\x00\x00\x0c\x00\x16\x00\x06\x00\x05\x00\x08\x00\x0c\x00\x0c\x00\x00\x00\x00\x03\x03\x00\x18\x00\x00\x000\x00\x00\x00\x00\x00\x00\x00\x00\x00\n\x00\x18\x00\x0c\x00\x04\x00\x08\x00\n\x00\x00\x00<\x00\x00\x00\x10\x00\x00\x00\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x000\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\xff\xff\xff\xff\x00\x00\x00\x00'
```

You can see the `0, 1, 2, 3, 4, 5` starting at byte 272: remember that it's 64-bits little endian, so it should look like this:

```python
>>> numpy.array([0, 1, 2, 3, 4, 5], dtype=numpy.int64).tostring()
b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00'
```

It's right near the end; I'll highlight it:

```python
>>> numpy.frombuffer(buf, dtype=numpy.uint8)[272:].tostring()
b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\xff\xff\xff\xff\x00\x00\x00\x00'
```

So now since the data buffers are identified by a pointer (from your copy of the C interface), I wonder if there are any `272` values in this? As a little-endian 64-bit integer, `272` looks like

```python
>>> numpy.array([272], dtype=numpy.int64).tostring()
b'\x10\x01\x00\x00\x00\x00\x00\x00'
```

Use Python to search for it:

```python
>>> numpy.frombuffer(buf, dtype=numpy.uint8).tostring().index(b'\x10\x01')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: subsection not found
```

No, so they must not mean a "pointer" as a relative index from the start of the buffer.

Incidentally, I see the name of our field, which must be part of the schema. It's at byte 124:

```python
>>> numpy.frombuffer(buf, dtype=numpy.uint8).tostring().index(b'one')
124
```

In other words, it's a reverse-engineering problem. Maybe the C structs don't map onto the buffer in exactly the way we expect—very likely, our interpretation of the text in that document is a little wrong. We can ask for clarifications and that will lead to more text, which we might still misinterpret. If the only tool at our disposal is a yes/no "does it fully deserialize with no errors?" then it will take a long time to converge, if ever. (More likely, they'll be forced to reduce it to exact instructions.) Or we can try to meet them halfway by looking at the data in one hand and the C structs in the other and thinking, "wait, what if what I thought started on byte zero actually starts on byte ten?" or something like that.

The `format` string in the C struct—what character should it be for 64-bit integers? Do we see that character in the buffer data anywhere? Since that's not indirected through a pointer, that would be a good hint of where the global offset is.

**Comment by jpivarski** [comment]
Oh, you have a fully working example going in the other direction: making a buffer from the C structs. That's great—make one of those for a single int64 array of `[0, 1, 2, 3, 4, 5]` and compare it with the bytes that pyarrow generates. Where are the

```python
b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00'
```

strings relative to one another? Is there more that's the same between them? Is it as simple as one being a subset of the other, maybe with a shift (that might be some kind of "envelope" that one or the other is wrapped up in)?

And you thought working in Python was going to be "high level."   `:)`

**Comment by trickarcher** [comment]
So I did follow the doc and created an `int64_t ArrowArray` "C" struct. There was a couple of things that didn't sit right with me:
1. The `ArrowSchema` is the only structure that differentiates different `ArrowArray` formats. However, the fact that I am not using it anywhere with the `ArrowArray` struct or for that matter for any kind of initialization which tells the Arrow library that "The next structure you will encounter would be of the kind that the `ArrowSchema` has provided you", doesn't seem correct to me.
2. I compared the `bytes` regardless and as expected they did not resemble any pattern. This behavior is expected because if you look closely you'll observe that the `buffers` are actually pointers and not the data themselves. You have to `dereference` the pointers and arrange them in a fashion that resembles an `ArrowArray`. This should imply (according to me) that there has to be some interface that makes sense of the C struct. The `ArrowArray` bytes are: `b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\xff\xff\xff\xff\x00\x00\x00\x00'` 
whereas the `C struct` bytes are:
`b'\n\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00p\xa0\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc3H\x9be\xad\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'`, which don't anything in common.

**Comment by jpivarski** [comment]
Okay, with this information, I think we can meaningfully ask for clarification on the Arrow GitHub Issues.

You have a lot of information here because you've done due diligence in trying to figure it out, but don't inundate them with this information or they won't know what the question is. We're probably going at it all wrong, and so explaining what we've tried (without showing big blocks of bytes) will be enough to prompt them to realize how we've gotten it wrong and what it is they need to clarify.

It should be clear what our goal is: to use the C interface with a minimum of dependencies to produce blocks of bytes that pyarrow can reconstruct as the same arrays (and vice-versa: both directions). We want to know what series of steps on the C side and what series of steps on the pyarrow side can be used to make a connection through nothing but a raw buffer of bytes (which might be in shared memory between processes, saved to a file, transmitted over a network, etc.).

In other words, the question should be short, make it clear what we need, and show what we've tried as a way to help them realize how their thinking is different from ours so that they can correct it. You should also link to the C interface document to help remind them what this is all about.

**Comment by trickarcher** [comment]
Yup, I'll do that in a bit.
